common/web/input-processor/src/text/prediction/languageProcessor.ts

```diff
@@ -167,6 +167,8 @@ export default class LanguageProcessor extends EventEmitter<LanguageProcessorEve
     } else if(outputTarget) {
       let transcription = outputTarget.buildTranscriptionFrom(outputTarget, null, false);
       return this.predict_internal(transcription, true, layerId);
+    } else {
+      return Promise.resolve([]);
```

Comment:
Is this a bug fix?

Reply:
`async` methods have an implicit `Promise` return.  `noImplicitReturns` throws compilation errors when this fact is relied upon for what would otherwise be a `void` return type.

Reply:
This is not an `asycn` function, so it was returning `undefined`, not `Promise<undefined>`. It is now returning `Promise<[]>`. That is a significant change (for the better) -- but we need to audit to make sure that all uses are correct!

In order to understand this more deeply, I did a small audit of the usage of `invalidateContext()`, and it was pretty complicated. 

It looks like it is used in two places, `InputProcessor.resetContext()` (which masks the fact that there is async going on by ignoring the returned promise), and `PredictionContext.resetContext()`.

`PredictionContext.setCurrentTarget()` returns the value from `PredictionContext.resetContext()`. The only other use of `PredictionContext.resetContext()` is in `ContextManagerBase.resetContext()`, and that ignores the return value (and again hides the async nature of the call). 

The return value from `PredictionContext.setCurrentTarget()` is not used in the KeymanWeb source. It is used only in 1 unit test, and that unit test would have failed if `undefined` had ever been returned.

This feels overly complicated and given we don't use the value outside unit test, perhaps we should just be returning `Promise<void>` from all those functions?

Reply:
If memory serves, the reason that it's like this is because that one unit test was _quite_ difficult to write otherwise.  A point of exposure was needed, and this fit best.

When all is said and done, the change here uses the following logic:  if there's no active context source, there's nothing to provide suggestions _for_.  In that case, there's no reason to even request suggestions, so bypass the prediction engine and say that there aren't any.

Reply:
I'd be happy to revisit this as a separate issue, but I don't think we want to add that sort of refactor within this PR.

Reply:
> if there's no active context source, there's nothing to provide suggestions for. In that case, there's no reason to even request suggestions, so bypass the prediction engine and say that there aren't any.

That'd be a great code comment here -- it says the 'why' here clearly, which saves a lot of spelunking!

```suggestion
    } else {
      // if there's no active context source, there's nothing to 
      // provide suggestions for. In that case, there's no reason 
      // to even request suggestions, so bypass the prediction 
      // engine and say that there aren't any.
      return Promise.resolve([]);
```

> I don't think we want to add that sort of refactor within this PR.

Agreed. The masked promises are an issue I think we need to try and address over time but not here.